src/app/pages/signature-request/signature-request.tsx

```diff
@@ -1,43 +1,59 @@
 
+import {
+  getSignaturePayloadFromToken,
+  getStructuredDataPayloadFromToken,
+} from '@app/common/signature/requests';
+import { isUndefined } from '@app/common/utils';
+import { openInNewTab } from '@app/common/utils/open-in-new-tab';
+import { Link } from '@app/components/link';
+import { Caption } from '@app/components/typography';
+import { PopupHeader } from '@app/features/current-account/popup-header';
 import {
   useIsSignatureRequestValid,
   useSignatureRequestSearchParams,
 } from '@app/store/signatures/requests.hooks';
-import { PageTop } from './components/page-top';
-import { MessageBox } from './components/message-box';
-import { NetworkRow } from './components/network-row';
 import {
   isSignatureMessageType,
-  SignAction,
-  SignatureMessage,
-  SignatureMessageType,
-} from './components/sign-action';
+  isStructuredMessage,
+  SignatureMessage as SignatureMessageProps,
+} from '@shared/signature/types';
 import { StacksNetwork, StacksTestnet } from '@stacks/network';
+import { ClarityValue } from '@stacks/transactions';
 import { FiAlertTriangle } from 'react-icons/fi';
-import { Caption } from '@app/components/typography';
-import { PopupHeader } from '@app/features/current-account/popup-header';
-import { getPayloadFromToken } from '@app/common/signature/requests';
-import { isUndefined } from '@app/common/utils';
-import { Link } from '@app/components/link';
-import { openInNewTab } from '@app/common/utils/open-in-new-tab';
+import { MessageBox } from './components/message-box';
+import { NetworkRow } from './components/network-row';
+import { PageTop } from './components/page-top';
+import { SignAction } from './components/sign-action';
+import { SignaturePayload, StructuredDataSignaturePayload } from '@stacks/connect';
 
 function SignatureRequestBase(): JSX.Element | null {
   const validSignatureRequest = useIsSignatureRequestValid();
   const { requestToken, messageType } = useSignatureRequestSearchParams();
 
   useRouteHeader(<PopupHeader />);
 
-  if (!requestToken || !messageType) return null;
-  const signatureRequest = getPayloadFromToken(requestToken);
-  if (!signatureRequest) return null;
+  if (!isSignatureMessageType(messageType)) return null;
   if (isUndefined(validSignatureRequest)) return null;
-  const appName = signatureRequest?.appDetails?.name;
-  const { message, network } = signatureRequest;
-  if (!isSignatureMessageType(message)) return null;
+  if (!requestToken || !messageType) return null;
+
+  let message: string | ClarityValue;
+  let signatureRequest: StructuredDataSignaturePayload | SignaturePayload;
+  let domain;
+  if (isStructuredMessage(messageType)) {
+    signatureRequest = getStructuredDataPayloadFromToken(requestToken);
+    domain = signatureRequest.domain;
+    message = signatureRequest.message;
+  } else {
+    signatureRequest = getSignaturePayloadFromToken(requestToken);
+    message = signatureRequest.message;
+  }
```

Comment:
Is there a cleaner way to do this, that doesn't involve conditional setting of empty vars? Should we be using two different components for string/structured message signing.

Reply:
Sure I will think of something else.